language-support/ts/daml-ledger/index.ts

```diff
@@ -584,13 +584,63 @@ class Ledger {
    * @typeparam T The contract template type.
    * @typeparam K The contract key type.
    * @typeparam I The contract id type.
+   *
+   * @deprecated since 1.5; use [[streamQueries]] instead
    */
+  // This code is deprecated and only kept to guarantee backwardss
+  // compatibility. Do not change until removal.
   streamQuery<T extends object, K, I extends string>(
     template: Template<T, K, I>,
     query?: Query<T>,
   ): Stream<T, K, I, readonly CreateEvent<T, K, I>[]> {
     const request = {templateIds: [template.templateId], query};
     const reconnectRequest = (): object[] => [request];
+    const change = (contracts: readonly CreateEvent<T, K, I>[], events: readonly Event<T, K, I>[]): CreateEvent<T, K, I>[] => {
+      const archiveEvents: Set<ContractId<T>> = new Set();
+      const createEvents: CreateEvent<T, K, I>[] = [];
+      for (const event of events) {
+        if ('created' in event) {
+          createEvents.push(event.created);
+        } else {
+          archiveEvents.add(event.archived.contractId);
+        }
+      }
+      return contracts
+        .concat(createEvents)
+        .filter(contract => !archiveEvents.has(contract.contractId));
+    };
+    return this.streamSubmit(template, 'v1/stream/query', request, reconnectRequest, [], change);
+  }
+
+  /**
+   * Retrieve a consolidated stream of events for a given template and queries.
+   *
+   * The accumulated state is the current set of active contracts matching the
+   * queries.
+   *
+   * When no `queries` argument is given, all events visible to the submitting
+   * party are returned.
+   *
+   * When one or more `queries` arguments are given, only those create events
+   * matching at least one of the queries are returned.
+   *
+   * See https://docs.daml.com/json-api/search-query-language.html for a
+   * description of the query language.
+   *
+   * @param template The contract template to match contracts against.
+   * @param query A query to match contracts against.
+   *
+   * @typeparam T The contract template type.
+   * @typeparam K The contract key type.
+   * @typeparam I The contract id type.
+   */
+  streamQueries<T extends object, K, I extends string>(
```

Comment:
That was also my impression at first. Based on the existing signature of `streamQuery`, I assumed existing invocations would look like either
```typescript
streamQuery(templateId)
```
or
```typescript
streamQuery(templateId, query)
```
and it felt like it would be very natural (and backwards compatible) to extend it to also support
```typescript
streamQuery(templateId, query1, query2)
```
Unfortunately, the current signature also allows
```typescript
streamQuery(templateId, undefined)
```
which means a backwards-compatible signature would have to look like
```typescript
streamQuery<T extends object, K, I extends string>(
     template: Template<T, K, I>,
     query?: Query<T>,
     ...queries: Query<T>[]
): ...
```
which in turn would force the resulting code to handle the case of
```typescript
streamQuery(templateId, undefined, query1, query2)
```
which just seemed wrong. So when I reached that point I decided deprecation was a better approach. I believe we are still at a stage where it is more important to have a nice API for new users than to not bother existing users, especially when there is a way to make the change through a backwards-compatible deprecation cycle. And I kinda like the new name better, for the new feature set.

That said, my knowledge of TypeScript is still very sparse at the moment, so I'm very open to the idea that there may be a way to preserve backwards compatibility _and_ get sane semantics for the new version and I'm just not aware of it.

Reply:
Good point! I forgot that in JS passing fewer parameters is basically equivalent to passing `undefined` (not entirely, you can differentiate them but almost nobody does). I see the following options:

1. Break the behavior on `streamQuery(templateId, undefined)` immediately. Technically a breaking change but I expect nobody to run into this and the fix is trivial.
2. Filter out `undefined` for the first query parameter or maybe even all query parameters. Should be easy to implement and makes this non-breaking.
3. 2 but simultaneously deprecate passing `undefined` (in that case, we should only filter it out from the first query). You can define multiple type signatures for one function in typescript. It might be possible to only deprecate one of them to get nice deprecation warnings. Alternatively, just throw a `console.log` at it.
4. The solution proposed in this PR.

I’m leaning towards 3 as probably the safest option (I expect nobody will actually have to change anything) and we can get rid of the wart later. But I’m happy to be convinced here.

Reply:
> But I’m happy to be convinced here.

Well then, allow me to try 🙂

I suspect "nobody" is actually using the raw `streamQuery`, instead of going through the [react bindings](https://github.com/digital-asset/daml/blob/master/language-support/ts/daml-react/createLedgerContext.ts#L122), which, by the way, do explicitly pass `undefined`. But it's still part of our public API so let's assume breaking it is a bad thing.

I believe option 1 is strictly inferior to this PR because it's a breaking change with no upgrade path: your code will just not compile if you try to upgrade. That seems bad. In comparison, this PR adds a new name and leaves the existing one completely unchanged. You can upgrade to the next version, then have some time to fix all the deprecation warnings while your app keeps working.

Options 2 and 3 seem very messy, and with the kind of mess that sticks. We'd be committing to an API with a singular name accepting a list of things that, according to their type, can be undefined, but with documentation saying they shouldn't be, and an implementation that makes them harmless for now. If we can indeed deprecate just one signature, we don't gain anything over this PR; if we can't, we'll end up with the downsides of 1 again later on. Or we'll live with the mess forever.

I don't quite understand why we'd shy away from a clean deprecation cycle, especially for a source-based language. However, there is another option you have not mentioned: keep `streamQuery` and `streamFetchByKey` exactly as they are right now, and add the two new functions with no deprecation. I believe that leaves our API overall a little bit confusing and clearly overlapping, but if you don't like deprecations I think that's the best approach.

Reply:
I’m not really worried about the deprecation cycle. That seems very reasonable. What I am slightly worried about is what seems like a fairly redundant name change. In this case, it somewhat makes sense but it still feels pretty arbitrary (the endpoint name is singular as well). Consider the case where we didn’t want to add more queries and just deprecate explicitly passing `undefined`. I have no idea what a sensible name would be there.

As for 3, we do gain something for the case where you explicitly pass a single query or don’t pass a query both of which need changing in this PR but not if we can only deprecate passing `undefined` explicitly.

Reply:
That said, your point that most people do not use this directly is pretty convincing. It does raise the question of what we do with the hooks though.

Reply:
I personally think changing the name is a good thing, as that is what allows us to have a clean deprecation cycle where both versions can be present in parallel and consumers of the API can very easily spot which places need to be updated. I agree a good second name would be harder to find if you just wanted to remove the `undefined` option, but even then I would argue for a new `streamQuery2` method over changing the existing one. Maybe I'm weird, though.

I think not having to change your code when the underlying behaviour has changed is actually a bad thing, so in my mind that counts against option 3.

I'll take a look at the react bindings. Agree they probably need some sort of similar update.

Reply:
Alright then let’s go with the new name here. I don’t mind tackling the react bindings in a separate PR.